File: cpp/test/behaviour/steps/common/transaction_steps.cpp
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <cstdlib>

#include "common.hpp"
#include "steps.hpp"
#include "utils.hpp"

using namespace cucumber::messages;

namespace TypeDB::BDD {

void forEachSessionTransaction_serial(Context& context, std::function<void(TypeDB::Transaction*)> fn) {
    for (auto& sess : context.sessions) {
        foreach_serial(context.sessionTransactions[&sess], fn);
    }
}

void forEachSessionTransaction_parallel(Context& context, std::function<void(TypeDB::Transaction*)> fn) {
    for (auto& sess : context.sessions) {
        foreach_parallel(context.sessionTransactions[&sess], fn);
    }
}

cucumber_bdd::StepCollection<Context> transactionSteps = {

    BDD_STEP("session opens transaction of type: (read|write)", {
        context.setTransaction(context.session().transaction(parseTransactionType(matches[1].str()), context.transactionOptions));
    }),
    BDD_UNIMPLEMENTED("session open transaction of type; throws exception: (read|write)"),
    BDD_UNIMPLEMENTED("for each session, open transactions of type; throws exception"),
    BDD_STEP("session transaction is null: (true|false)", {
        ASSERT_EQ(parseBoolean(matches[1]), !context.transaction().isOpen());
    }),
    BDD_STEP("session transaction is open: (true|false)", {
        ASSERT_EQ(parseBoolean(matches[1]), context.transaction().isOpen());
    }),
    BDD_STEP("(?:session )?transaction commits", {
        context.transaction().commit();
    }),
    BDD_STEP("session transaction closes", {
        context.transaction().close();
    }),
    BDD_STEP("(?:session )?transaction commits; throws exception", {
        DRIVER_THROWS("", { context.transaction().commit(); });
    }),
    BDD_STEP("transaction commits; throws exception containing \"(.+)\"", {
        DRIVER_THROWS(matches[1].str(), { context.transaction().commit(); });
    }),
    BDD_STEP("session transaction has type: (read|write)", {
        ASSERT_EQ(parseTransactionType(matches[1].str()), context.transaction().type());
    }),
    BDD_STEP("set transaction option ([A-Za-z_\\-]+) to: ([A-Za-z0-9]+)", {
        assert(matches[1] == "transaction-timeout-millis");
        context.transactionOptions.transactionTimeoutMillis(atoi(matches[2].str().c_str()));
    }),

    BDD_STEP("for each session, open transaction of type: (read|write)", {
        std::function<void(TypeDB::Session*)> fn = [&](TypeDB::Session* session) {
            context.sessionTransactions[session].push_back(session->transaction(parseTransactionType(matches[1].str()), context.transactionOptions));
        };
        foreach_serial(context.sessions, fn);
    }),
    BDD_UNIMPLEMENTED("for each session, open transaction of type:"),
    BDD_STEP("for each session, open transactions of type:", {
        std::function<void(TypeDB::Session*)> fn = [&](TypeDB::Session* session) {
            for (auto& row : step.argument->data_table->rows) {
                context.sessionTransactions[session].push_back(session->transaction(parseTransactionType(row.cells[0].value), context.transactionOptions));
            }
        };
        foreach_serial(context.sessions, fn);
    }),
    BDD_STEP("for each session, transaction[s]? (?:are|is) null: (true|false)", {  // overload
        std::function<void(TypeDB::Transaction*)> fn = [&](TypeDB::Transaction* tx) { ASSERT_EQ(parseBoolean(matches[1]), !tx->isOpen()); };
        forEachSessionTransaction_serial(context, fn);
    }),
    BDD_STEP("for each session, transaction[s]? (?:are|is) open: (true|false)", {  // overload
        std::function<void(TypeDB::Transaction*)> fn = [&](TypeDB::Transaction* tx) -> void { ASSERT_EQ(parseBoolean(matches[1]), tx->isOpen()); };
        forEachSessionTransaction_serial(context, fn);
    }),
    BDD_STEP("for each session, transaction[s]? commit[s]?", {  // overload
        std::function<void(TypeDB::Transaction*)> fn = [&](TypeDB::Transaction* transaction) { transaction->commit(); };
        forEachSessionTransaction_serial(context, fn);
    }),
    BDD_STEP("for each session, transaction commit[s]?; throws exception", {  // overload
        std::function<void(TypeDB::Transaction*)> fn = [&](TypeDB::Transaction* transaction) { DRIVER_THROWS("", transaction->commit()); };
        forEachSessionTransaction_serial(context, fn);
    }),
    BDD_STEP("for each session, transaction closes", {
        std::function<void(TypeDB::Transaction*)> fn = [&](TypeDB::Transaction* transaction) { transaction->close(); };
        forEachSessionTransaction_serial(context, fn);
    }),
    BDD_STEP("for each session, transaction has type", {
        std::vector<TypeDB::Transaction*> transactions;
        for (auto& session : context.sessions) {
            for (auto& txn : context.sessionTransactions[&session]) {
                transactions.push_back(&txn);
            }
        }
        ASSERT_EQ(step.argument->data_table->rows.size(), transactions.size());
        for (int i = 0; i < transactions.size(); i++) {
            ASSERT_EQ(parseTransactionType(step.argument->data_table->rows[i].cells[0].value), transactions[i]->type());
        }
    }),
    BDD_STEP("for each session, transactions have type:", {
        std::vector<TypeDB::Transaction*> transactions;
        for (auto& session : context.sessions) {
            ASSERT_EQ(step.argument->data_table->rows.size(), context.sessionTransactions[&session].size());
            for (int i = 0; i < step.argument->data_table->rows.size(); i++) {
                ASSERT_EQ(parseTransactionType(step.argument->data_table->rows[i].cells[0].value), context.sessionTransactions[&session][i].type());
            }
        }
    }),
    BDD_STEP("for each session, transaction has type: (read|write)", {
        std::function<void(TypeDB::Transaction*)> fn = [&](TypeDB::Transaction* transaction) { ASSERT_EQ(parseTransactionType(matches[1].str()), transaction->type()); };
        forEachSessionTransaction_serial(context, fn);
    }),

    BDD_STEP("for each session, open transactions in parallel of type:", {
        std::vector<int> nums;
        for (int i = 0; i < step.argument->data_table->rows.size(); i++)
            nums.push_back(i);

        for (TypeDB::Session& session : context.sessions) {
            std::vector<std::optional<TypeDB::Transaction>> txn(nums.size());
            std::function<void(int*)> fn = [&](int* i) {
                txn[*i] = session.transaction(parseTransactionType(step.argument->data_table->rows[*i].cells[0].value), context.transactionOptions);
            };

            foreach_parallel(nums, fn);

            for (auto& tx : txn) {
                context.sessionTransactions[&session].push_back(std::move(tx.value()));
            }
        }
    }),
    BDD_STEP("for each session, transactions in parallel are null: (true|false)", {
        std::function<void(TypeDB::Transaction*)> fn = [&](TypeDB::Transaction* tx) { ASSERT_EQ(parseBoolean(matches[1]), !tx->isOpen()); };
        forEachSessionTransaction_parallel(context, fn);
    }),
    BDD_STEP("for each session, transactions in parallel are open: (true|false)", {
        std::function<void(TypeDB::Transaction*)> fn = [&](TypeDB::Transaction* tx) { ASSERT_EQ(parseBoolean(matches[1]), tx->isOpen()); };
        forEachSessionTransaction_parallel(context, fn);
    }),
    BDD_STEP("for each session, transactions in parallel have type:", {
        std::function<void(zipped<Transaction>*)> fn = [&](zipped<Transaction>* zipped) {
            ASSERT_EQ(parseTransactionType(zipped->row->cells[0].value), zipped->obj->type());
        };
        for (auto& session : context.sessions) {
            auto z = zip(step.argument->data_table->rows, context.sessionTransactions[&session]);
            foreach_parallel(z, fn);
        }
    }),

    // I don't see them in the BDD
    BDD_UNIMPLEMENTED("for each transaction, define query; throws exception containing \"(.+)\""),
    BDD_UNIMPLEMENTED("for each session in parallel, transactions in parallel are null: (true|false)"),
    BDD_UNIMPLEMENTED("for each session in parallel, transactions in parallel are open: (true|false)"),

};

}  // namespace TypeDB::BDD

File: python/tests/behaviour/connection/transaction/transaction_steps.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from concurrent.futures.thread import ThreadPoolExecutor
from functools import partial
from typing import Callable, Optional

from behave import *
from hamcrest import *
from tests.behaviour.config.parameters import MayError, parse_list, parse_transaction_type
from tests.behaviour.context import Context
from typedb.api.connection.transaction import TransactionType
from typedb.driver import *


def open_transactions_of_type(context: Context, database_name: str, transaction_types: list[TransactionType]):
    context.transactions = []
    for transaction_type in transaction_types:
        transaction = context.driver.transaction(database_name, transaction_type)  # , context.transaction_options
        context.transactions.append(transaction)


@step(
    "connection open {transaction_type:TransactionType} transaction for database: {database_name:Words}{may_error:MayError}")
def step_impl(context: Context, transaction_type: TransactionType, database_name: str, may_error: MayError):
    may_error.check(lambda: open_transactions_of_type(context, database_name, [transaction_type]))


@step("connection open transactions for database: {database_name}, of type")
def step_impl(context: Context, database_name: str):
    transaction_types = list(map(parse_transaction_type, parse_list(context.table)))
    open_transactions_of_type(context, database_name, transaction_types)


def for_each_transaction_is(context: Context, assertion: Callable[[Transaction], None]):
    for transaction in context.transactions:
        assertion(transaction)


def assert_transaction_open(transaction: Optional[Transaction], is_open: bool):
    assert_that(transaction is not None and transaction.is_open(), is_(is_open))


@step("transaction is open: {is_open:Bool}")
def step_impl(context: Context, is_open: bool):
    assert_transaction_open(context.tx(), is_open)


@step("transactions are open: {are_open:Bool}")
def step_impl(context: Context, are_open: bool):
    for_each_transaction_is(context, lambda tx: assert_transaction_open(tx, are_open))


@step("transaction commits{may_error:MayError}")
def step_impl(context: Context, may_error: MayError):
    may_error.check(lambda: context.tx().commit())


@step("transaction closes")
def step_impl(context: Context):
    context.tx().close()


@step("transaction rollbacks{may_error:MayError}")
def step_impl(context: Context, may_error: MayError):
    may_error.check(lambda: context.tx().rollback())


def for_each_transaction_has_type(context: Context, transaction_types: list):
    assert_that(context.transactions, has_length(len(transaction_types)))
    transactions_iterator = iter(context.transactions)
    for transaction_type in transaction_types:
        assert_that(next(transactions_iterator).type, is_(transaction_type))


@step("transactions have type")
def step_impl(context: Context):
    transaction_types = list(map(parse_transaction_type, parse_list(context.table)))
    for_each_transaction_has_type(context, transaction_types)


@step("transaction has type: {transaction_type:TransactionType}")
def step_impl(context: Context, transaction_type: TransactionType):
    for_each_transaction_has_type(context, [transaction_type])


@step("connection open transactions in parallel for database: {name:Words}, of type")
def step_impl(context: Context, name: str):
    types = list(map(parse_transaction_type, parse_list(context.table)))
    assert_that(len(types), is_(less_than_or_equal_to(context.THREAD_POOL_SIZE)))
    with ThreadPoolExecutor(max_workers=context.THREAD_POOL_SIZE) as executor:
        context.transactions_parallel = []
        for type_ in types:
            context.transactions_parallel.append(
                executor.submit(partial(context.driver.transaction, name, type_)))


def transactions_in_parallel_are(context: Context, assertion: Callable[[Transaction], None]):
    for future_transaction in context.transactions_parallel:
        assertion(future_transaction.result())


@step("transactions in parallel are open: {are_open:Bool}")
def step_impl(context: Context, are_open: bool):
    transactions_in_parallel_are(context, lambda tx: assert_transaction_open(tx, are_open))


@step("transactions in parallel have type")
def step_impl(context: Context):
    types = list(map(parse_transaction_type, parse_list(context.table)))
    future_transactions = context.transactions_parallel
    assert_that(future_transactions, has_length(len(types)))
    future_transactions_iter = iter(future_transactions)
    for type_ in types:
        assert_that(next(future_transactions_iter).result().type, is_(type_))


@step("set transaction option {option} to: {value:Int}")
def step_impl(context: Context, option: str, value: int):
    if option not in context.option_setters:
        raise Exception("Unrecognised option: " + option)
    context.option_setters[option](context.transaction_options, value)


@step("schedule database creation on transaction close: {database_name:Words}")
def step_impl(context: Context, database_name: str):
    context.tx().on_close(lambda: context.driver.databases.create(database_name))
